```
=== RUN   TestValidatorOverwriteEmailListViaCopyingOver
2018/11/23 23:13:44 validator.go:23: using authenticated emails file /tmp/test_auth_emails_108908214
2018/11/23 23:13:44 watcher.go:67: watching /tmp/test_auth_emails_108908214 for updates
2018/11/23 23:13:44 watcher.go:53: watching interrupted on event: "/tmp/test_auth_emails_108908214": CHMOD
2018/11/23 23:13:44 watcher.go:25: watching resumed for /tmp/test_auth_emails_108908214
2018/11/23 23:13:44 watcher.go:57: reloading after event: "/tmp/test_auth_emails_108908214": CHMOD
2018/11/23 23:13:44 watcher.go:53: watching interrupted on event: "/tmp/test_auth_emails_108908214": REMOVE
2018/11/23 23:13:44 watcher.go:25: watching resumed for /tmp/test_auth_emails_108908214
2018/11/23 23:13:44 watcher.go:57: reloading after event: "/tmp/test_auth_emails_108908214": REMOVE
2018/11/23 23:13:44 validator.go:42: failed opening authenticated-emails-file="/tmp/test_auth_emails_108908214", open /tmp/test_auth_emails_108908214: no such file or directory
FAIL	github.com/bitly/oauth2_proxy	4.680s
```

```
=== RUN   TestValidatorOverwriteEmailListViaCopyingOver
2018/11/23 23:14:02 watcher.go:53: watching interrupted on event: "/tmp/test_auth_emails_452869335": REMOVE
2018/11/23 23:14:02 validator.go:23: using authenticated emails file /tmp/test_auth_emails_173157450
2018/11/23 23:14:02 watcher.go:67: watching /tmp/test_auth_emails_173157450 for updates
2018/11/23 23:14:02 watcher.go:53: watching interrupted on event: "/tmp/test_auth_emails_173157450": CHMOD
panic: test timed out after 1m30s
goroutine 215 [running]:
testing.(*M).startAlarm.func1()
	/home/travis/.gimme/versions/go1.11.2.linux.amd64/src/testing/testing.go:1296 +0x13a
created by time.goFunc
	/home/travis/.gimme/versions/go1.11.2.linux.amd64/src/time/sleep.go:172 +0x52

...

goroutine 243 [chan receive]:
github.com/bitly/oauth2_proxy.TestValidatorOverwriteEmailListViaCopyingOver(0xc000391000)
	/home/travis/gopath/src/github.com/bitly/oauth2_proxy/validator_watcher_copy_test.go:43 +0x1ad
testing.tRunner(0xc000391000, 0xbd06b8)
	/home/travis/.gimme/versions/go1.11.2.linux.amd64/src/testing/testing.go:827 +0x163
created by testing.(*T).Run
	/home/travis/.gimme/versions/go1.11.2.linux.amd64/src/testing/testing.go:878 +0x651

goroutine 239 [sleep]:
time.Sleep(0x2faf080)
	/home/travis/.gimme/versions/go1.11.2.linux.amd64/src/runtime/time.go:105 +0x14f
github.com/bitly/oauth2_proxy.WaitForReplacement(0xc00038c700, 0x1f, 0x4, 0xc0003e6190)
	/home/travis/gopath/src/github.com/bitly/oauth2_proxy/watcher.go:29 +0x4b
github.com/bitly/oauth2_proxy.WatchForUpdates.func1(0xc0003e6190, 0xc000039960, 0xc00038c700, 0x1f, 0xc000374e60)
	/home/travis/gopath/src/github.com/bitly/oauth2_proxy/watcher.go:55 +0x3a2
created by github.com/bitly/oauth2_proxy.WatchForUpdates
	/home/travis/gopath/src/github.com/bitly/oauth2_proxy/watcher.go:39 +0x1ee

goroutine 238 [syscall]:
syscall.Syscall6(0xe8, 0x17, 0xc0003b2cdc, 0x7, 0xffffffffffffffff, 0x0, 0x0, 0x465e66, 0x4119f0, 0x2564528)
	/home/travis/.gimme/versions/go1.11.2.linux.amd64/src/syscall/asm_linux_amd64.s:44 +0x5
syscall.EpollWait(0x17, 0xc0003b2cdc, 0x7, 0x7, 0xffffffffffffffff, 0xc0003b2ca8, 0x48db0e, 0xc0003e61a0)
	/home/travis/.gimme/versions/go1.11.2.linux.amd64/src/syscall/zsyscall_linux_amd64.go:1714 +0x86
github.com/bitly/oauth2_proxy/vendor/github.com/fsnotify/fsnotify.(*fdPoller).wait(0xc00038c760, 0xc0003e6100, 0x800000000001, 0xc00036bf01)
	/home/travis/gopath/src/github.com/bitly/oauth2_proxy/vendor/github.com/fsnotify/fsnotify/inotify_poller.go:85 +0xb9
github.com/bitly/oauth2_proxy/vendor/github.com/fsnotify/fsnotify.(*Watcher).readEvents(0xc0003e6190)
	/home/travis/gopath/src/github.com/bitly/oauth2_proxy/vendor/github.com/fsnotify/fsnotify/inotify.go:191 +0x24f
created by github.com/bitly/oauth2_proxy/vendor/github.com/fsnotify/fsnotify.NewWatcher
	/home/travis/gopath/src/github.com/bitly/oauth2_proxy/vendor/github.com/fsnotify/fsnotify/inotify.go:60 +0x3ce

... a bunch more coroutines sleeping in WaitForReplacement or inotify epoll

goroutine 184 [sync.Cond.Wait]:
sync.runtime_notifyListWait(0xc0001d8dd0, 0x0)
	/home/travis/.gimme/versions/go1.11.2.linux.amd64/src/runtime/sema.go:510 +0xeb
sync.(*Cond).Wait(0xc0001d8dc0)
	/home/travis/.gimme/versions/go1.11.2.linux.amd64/src/sync/cond.go:56 +0x8e
github.com/bitly/oauth2_proxy/vendor/github.com/fsnotify/fsnotify.(*Watcher).Remove(0xc00020ec80, 0xc00038ca40, 0x1f, 0x0, 0x0)
	/home/travis/gopath/src/github.com/bitly/oauth2_proxy/vendor/github.com/fsnotify/fsnotify/inotify.go:157 +0x1f1
github.com/bitly/oauth2_proxy.WatchForUpdates.func1(0xc00020ec80, 0xc000270850, 0xc00038ca40, 0x1f, 0xc0001f67c0)
	/home/travis/gopath/src/github.com/bitly/oauth2_proxy/watcher.go:54 +0x36f
created by github.com/bitly/oauth2_proxy.WatchForUpdates
	/home/travis/gopath/src/github.com/bitly/oauth2_proxy/watcher.go:39 +0x1ee

goroutine 183 [select]:
github.com/bitly/oauth2_proxy/vendor/github.com/fsnotify/fsnotify.(*Watcher).readEvents(0xc00020ec80)
	/home/travis/gopath/src/github.com/bitly/oauth2_proxy/vendor/github.com/fsnotify/fsnotify/inotify.go:265 +0x781
created by github.com/bitly/oauth2_proxy/vendor/github.com/fsnotify/fsnotify.NewWatcher
	/home/travis/gopath/src/github.com/bitly/oauth2_proxy/vendor/github.com/fsnotify/fsnotify/inotify.go:60 +0x3ce
```

--- a/validator_test.go
+++ b/validator_test.go
@@ -20,7 +20,8 @@ func NewValidatorTest(t *testing.T) *ValidatorTest {
 	if err != nil {
 		t.Fatal("failed to create temp file: " + err.Error())
 	}
-	vt.done = make(chan bool, 1)
+	t.Logf("%s using temp file %s", t.Name(), vt.auth_email_file.Name())
+	vt.done = make(chan bool)
 	return vt
 }
 
@@ -40,7 +41,6 @@ func (vt *ValidatorTest) NewValidator(domains []string, updated chan<- bool) fun
 
 // This will close vt.auth_email_file.
 func (vt *ValidatorTest) WriteEmails(t *testing.T, emails []string) {
-	defer vt.auth_email_file.Close()
 	vt.auth_email_file.WriteString(strings.Join(emails, "\n"))
 	if err := vt.auth_email_file.Close(); err != nil {
 		t.Fatal("failed to close temp file " +

--- a/validator_watcher_copy_test.go
+++ b/validator_watcher_copy_test.go
@@ -32,7 +32,7 @@ func TestValidatorOverwriteEmailListViaCopyingOver(t *testing.T) {
 
 	vt.WriteEmails(t, []string{"xyzzy@example.com"})
 	domains := []string(nil)
-	updated := make(chan bool)
+	updated := make(chan bool, 1)
 	validator := vt.NewValidator(domains, updated)
 
 	if !validator("xyzzy@example.com") {

--- a/validator_watcher_test.go
+++ b/validator_watcher_test.go
@@ -49,7 +49,7 @@ func TestValidatorOverwriteEmailListDirectly(t *testing.T) {
 		"plugh@example.com",
 	})
 	domains := []string(nil)
-	updated := make(chan bool)
+	updated := make(chan bool, 1)
 	validator := vt.NewValidator(domains, updated)
 
 	if !validator("xyzzy@example.com") {

--- a/watcher.go
+++ b/watcher.go
@@ -11,23 +11,18 @@ import (
 	"github.com/fsnotify/fsnotify"
 )
 
-func WaitForReplacement(filename string, op fsnotify.Op,
-	watcher *fsnotify.Watcher) {
-	const sleep_interval = 50 * time.Millisecond
+func WaitForReplacement(filename string, watcher *fsnotify.Watcher) {
+	for i := 0; i < 20; i++ {
+		time.Sleep(100 * time.Millisecond)
 
-	// Avoid a race when fsnofity.Remove is preceded by fsnotify.Chmod.
-	if op&fsnotify.Chmod != 0 {
-		time.Sleep(sleep_interval)
-	}
-	for {
 		if _, err := os.Stat(filename); err == nil {
 			if err := watcher.Add(filename); err == nil {
 				log.Printf("watching resumed for %s", filename)
 				return
 			}
 		}
-		time.Sleep(sleep_interval)
 	}
+	log.Printf("failed to resume watching for %s", filename)
 }
 
 func WatchForUpdates(filename string, done <-chan bool, action func()) {
@@ -37,22 +32,22 @@ func WatchForUpdates(filename string, done <-chan bool, action func()) {
 		log.Fatal("failed to create watcher for ", filename, ": ", err)
 	}
 	go func() {
-		defer watcher.Close()
 		for {
 			select {
 			case _ = <-done:
 				log.Printf("Shutting down watcher for: %s", filename)
-				break
+				watcher.Close()
+				return
 			case event := <-watcher.Events:
 				// On Arch Linux, it appears Chmod events precede Remove events,
 				// which causes a race between action() and the coming Remove event.
-				// If the Remove wins, the action() (which calls
-				// UserMap.LoadAuthenticatedEmailsFile()) crashes when the file
-				// can't be opened.
-				if event.Op&(fsnotify.Remove|fsnotify.Rename|fsnotify.Chmod) != 0 {
+				if event.Op == fsnotify.Chmod {
+					continue
+				}
+				if event.Op&(fsnotify.Remove|fsnotify.Rename) != 0 {
 					log.Printf("watching interrupted on event: %s", event)
 					watcher.Remove(filename)
-					WaitForReplacement(filename, event.Op, watcher)
+					WaitForReplacement(filename, watcher)
 				}
 				log.Printf("reloading after event: %s", event)
 				action()

